### PR TITLE
Add end() to Generation and Span

### PR DESCRIPTION
**Pull request by Dev-Khant**

### CODE REVIEW

**Comment by Dev-Khant** [comment]
@marcklingen @maxdeichmann Let me know of any required changes.

File: langfuse/client.py [hunk 1]
```diff
@@ -138,8 +139,7 @@ def create_trace():
 
             def create_span():
                 try:
-                    new_body = body.copy(update={"id": new_span_id})
-                    new_body = body.copy(update={"trace_id": new_trace_id})
+                    new_body = body.copy(update={"id": new_span_id, "trace_id": new_trace_id})
```
Comment (Dev-Khant) [inline langfuse/client.py:142]: Reason : 2 seperate spans were created from same trace because before this change when setting `trace_id` , `id` was set to None.

File: langfuse/client.py [hunk 2]
```diff
@@ -179,8 +179,7 @@ def create_trace():
 
             def create_generation():
                 try:
-                    new_body = body.copy(update={"id": new_generation_id})
-                    new_body = body.copy(update={"trace_id": new_trace_id})
+                    new_body = body.copy(update={"id": new_generation_id, "trace_id": new_trace_id})
```
Comment (Dev-Khant) [inline langfuse/client.py:182]: Reason : 2 seperate generations were created from same trace because before this change when setting `trace_id` , `id` was set to None.

File: langfuse/api/resources/span/types/update_span_request.py [hunk 3]
```diff
@@ -12,6 +12,7 @@
 class UpdateSpanRequest(pydantic.BaseModel):
     span_id: str = pydantic.Field(alias="spanId")
     trace_id: typing.Optional[str] = pydantic.Field(alias="traceId")
+    name: typing.Optional[str]
```
Comment (maxdeichmann) [inline langfuse/api/resources/span/types/update_span_request.py:15]: Good catch! All files in langfuse/api are auto-generated by fern. Added a PR on the core project and will update this PR as well https://github.com/langfuse/langfuse/pull/101

By keeping the openapi docs in core in sync, we can also update the TS SDK accordingly.

File: langfuse/client.py (hunk 1 above)
Comment (maxdeichmann) [reply langfuse/client.py:142]: good catch!

File: langfuse/client.py (hunk 2 above)
Comment (maxdeichmann) [reply langfuse/client.py:182]: same here!

File: langfuse/client.py [hunk 4]
```diff
@@ -190,8 +189,8 @@ def create_generation():
                     self.log.exception(e)
                     raise e
 
-            self.task_manager.add_task(new_generation_id, create_generation)
             self.task_manager.add_task(new_trace_id, create_trace)
+            self.task_manager.add_task(new_generation_id, create_generation)
```
Comment (maxdeichmann) [inline langfuse/client.py:193]: Ordering does not matter, but more logical to have it this way

File: langfuse/client.py [hunk 5]
```diff
@@ -358,7 +357,7 @@ def update(self, body: UpdateGeneration):
 
             def task():
                 try:
-                    new_body = body.copy(update={"generation_id": self.id})
+                    new_body = body.copy(update={"generation_id": self.id, "trace_id": self.trace_id})
```
Comment (maxdeichmann) [inline langfuse/client.py:360]: Why this change? It does not destroy anything but did it not work without?

File: tests/test_sdk.py [hunk 6]
```diff
@@ -424,3 +426,76 @@ def test_create_trace_with_id_and_generation():
...
+
+    assert len(langfuse.task_manager.result_mapping) == 3
+
+    trace_id = langfuse.get_trace_id()
+
+    trace = api_wrapper.get_trace(trace_id)
+
+    span = trace["observations"][0]
+    iso_format_end_time = end_time.astimezone(pytz.UTC).strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3] + "Z"
+    assert span["endTime"] == iso_format_end_time
```
Comment (maxdeichmann) [inline tests/test_sdk.py:469]: Do you think these type of tests work reliably as they base on time? Maybe we should just check that the endtime is not null?

Comment (Dev-Khant) [reply tests/test_sdk.py:469]: I thought on the platform time is shown in UTC time zone. But yes also checking if it's null or not is better. I'll make the change.

File: langfuse/client.py (hunk 5 above)
Comment (Dev-Khant) [reply langfuse/client.py:360]: Before this change it was not actually updating the generation because it couldn't find which trace it is.

File: langfuse/client.py (hunk 4 above)
Comment (Dev-Khant) [reply langfuse/client.py:193]: Yes

File: langfuse/api/resources/span/types/update_span_request.py (hunk 3 above)
Comment (Dev-Khant) [reply langfuse/api/resources/span/types/update_span_request.py:15]: Alright

**Comment by Dev-Khant** [comment]
@maxdeichmann I made the change to check for Null for endTime.

File: langfuse/client.py [hunk 7]
```diff
@@ -401,6 +416,14 @@ def task():
         except Exception as e:
             self.log.exception(e)
 
+    def end(self):
+        try:
+            end_time = datetime.now()
+            self.update(UpdateGeneration(endTime=end_time))
```
Comment (maxdeichmann) [inline langfuse/client.py:422]: I missed this one here: i think we need to return this function here so that end() returns a stateful client (also above). Otherwise good for me.

**Review by maxdeichmann — APPROVED** [review]
